File: backend/app/schemas/device.py
"""
设备相关的Pydantic模式

定义设备管理相关的数据结构
"""

from pydantic import BaseModel, HttpUrl
from typing import Optional, List
from datetime import datetime


class DeviceBase(BaseModel):
    """设备基础模式"""
    name: Optional[str] = None
    device_type: Optional[str] = None
    api_base_url: str
    business_node_id: Optional[str] = None


class DeviceCreate(BaseModel):
    """设备创建模式"""
    api_base_url: str
    username: str
    password: str


class DeviceUpdate(BaseModel):
    """设备更新模式"""
    name: Optional[str] = None
    device_type: Optional[str] = None
    api_base_url: Optional[str] = None
    business_node_id: Optional[str] = None
    status: Optional[str] = None


class DeviceResponse(DeviceBase):
    """设备响应模式"""
    id: int
    status: str
    device_sn: Optional[str] = None
    version: Optional[str] = None
    register_status: Optional[str] = None
    system_update: Optional[str] = None
    system_time: Optional[str] = None
    network_setting: Optional[str] = None
    last_sync_at: Optional[datetime] = None
    last_heartbeat: Optional[datetime] = None
    sync_error: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class DeviceListResponse(BaseModel):
    """设备列表响应模式"""
    items: List[DeviceResponse]
    total: int
    page: int = 1
    page_size: int = 50


class DeviceSync(BaseModel):
    """设备同步模式"""
    device_ids: List[int]


class DeviceBatchImport(BaseModel):
    """批量导入设备模式"""
    devices: List[DeviceCreate]


class DeviceImportResult(BaseModel):
    """设备导入结果"""
    row: int
    success: bool
    message: str
    device_id: Optional[int] = None


class DeviceBatchImportResponse(BaseModel):
    """批量导入响应模式"""
    success: bool
    message: str
    success_count: int
    failure_count: int
    results: List[DeviceImportResult]
    failures: List[dict] = []


class DeviceSyncResult(BaseModel):
    """设备同步结果"""
    device_id: int
    success: bool
    message: str
    data: Optional[dict] = None


class DeviceBatchSyncResponse(BaseModel):
    """批量同步响应模式"""
    success: bool
    message: str
    results: List[DeviceSyncResult]
    success_count: int
    failure_count: int


class DeviceStats(BaseModel):
    """设备统计模式"""
    total: int
    online: int
    offline: int
    error: int
    unknown: int


# 设备API相关模式（调用外部设备时使用）
class DeviceLoginRequest(BaseModel):
    """设备登录请求"""
    username: str
    password: str


class DeviceLoginResponse(BaseModel):
    """设备登录响应"""
    token: str
    user: dict


class DeviceInfoResponse(BaseModel):
    """设备信息响应"""
    device_name: Optional[str] = None
    device_type: Optional[str] = None
    device_sn: Optional[str] = None
    software_version: Optional[str] = None
    register_status: Optional[str] = None
    system_update: Optional[str] = None
    system_time: Optional[str] = None
    network_setting: Optional[str] = None

File: frontend_ant/src/api/device.ts
/**
 * 设备管理 API
 */

import { request } from './request'

export interface Device {
  id: number
  name?: string
  device_type?: string
  api_base_url: string
  status: 'online' | 'offline' | 'error'
  business_node_id?: string
  device_sn?: string
  version?: string
  register_status?: string
  system_update?: string
  system_time?: string
  network_setting?: string
  last_sync_at?: string
  last_heartbeat?: string
  sync_error?: string
  created_at?: string
  updated_at?: string
}

export interface DeviceCreate {
  name?: string
  api_base_url: string
  username: string
  password: string
}

export interface DeviceListResponse {
  success: boolean
  data: {
    items: Device[]
    total: number
    page: number
    page_size: number
  }
}

export interface DeviceResponse {
  success: boolean
  data: Device
}

export interface DeviceStats {
  total: number
  online: number
  offline: number
  error: number
}

export interface DeviceSyncResult {
  success: boolean
  message: string
  data?: any
}

export interface DeviceBatchSyncResponse {
  success: boolean
  message: string
  results: Array<{
    device_id: number
    success: boolean
    message: string
  }>
}

export interface DeviceBatchImport {
  devices: DeviceCreate[]
}

export interface DeviceBatchImportResponse {
  success: boolean
  message: string
  imported_count: number
  failed_count: number
  failed_devices: Array<{
    device: DeviceCreate
    error: string
  }>
}

export interface DeviceSource {
  id: number
  name: string
  source_type: string
  url: string
  status: string
  resolution?: string
  thumbnail_url?: string
  description?: string
}

// 获取设备列表
export const getDevices = (params: {
  page?: number
  page_size?: number
  status?: string
  device_type?: string
  search?: string
} = {}) => {
  return request.get<DeviceListResponse>('/devices/', { params })
}

// 获取设备详情
export const getDevice = (deviceId: number) => {
  return request.get<DeviceResponse>(`/devices/${deviceId}`)
}

// 添加设备
export const addDevice = (data: DeviceCreate) => {
  return request.post<DeviceResponse>('/devices/', data)
}

// 更新设备
export const updateDevice = (deviceId: number, data: DeviceCreate) => {
  return request.put<DeviceResponse>(`/devices/${deviceId}`, data)
}

// 删除设备
export const deleteDevice = (deviceId: number) => {
  return request.delete(`/devices/${deviceId}`)
}

// 同步设备信息
export const syncDevice = (deviceId: number) => {
  return request.post<DeviceSyncResult>(`/devices/${deviceId}/sync`)
}

// 批量同步设备
export const batchSyncDevices = (deviceIds: number[]) => {
  return request.post<DeviceBatchSyncResponse>('/devices/batch-sync', { device_ids: deviceIds })
}

// 批量删除设备
export const batchDeleteDevices = (deviceIds: number[]) => {
  return request.delete('/devices/batch-delete', { data: { device_ids: deviceIds } })
}

// 批量导入设备
export const batchImportDevices = (devices: DeviceCreate[]) => {
  return request.post<DeviceBatchImportResponse>('/devices/batch-import', { devices })
}

// 获取设备统计
export const getDeviceStats = () => {
  return request.get<DeviceStats>('/devices/stats/overview')
}

// 获取设备视频源
export const getDeviceSources = (deviceId: number, sourceType?: string) => {
  const params = sourceType ? { source_type: sourceType } : {}
  return request.get<{
    success: boolean
    data: DeviceSource[]
    total: number
  }>(`/devices/${deviceId}/sources`, { params })
}